llvm/include/llvm/SYCLLowerIR/ESIMD/LowerESIMD.h

```diff
@@ -80,6 +80,12 @@ class SYCLFixupESIMDKernelWrapperMDPass
 public:
   PreservedAnalyses run(Module &M, ModuleAnalysisManager &);
 };
+
+class ESIMDRemoveHostCodePass : public PassInfoMixin<ESIMDRemoveHostCodePass> {
```

Comment:
Wouldn't it be even better removing host code earlier - before that code passed through all optimizations.
For example, completely clear the kernel function here: https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaSYCL.cpp#L4023

Reply:
So first, I made sure the new pass runs pretty early. In the testing I just did, it is literally the first pass run after LLVM IR is generated. And the problem isn't really the size of the LLVM IR, it's the size (and time it takes) to generate the required X86 asm, so I don't think skipping the Clang LLVM IR codegen would give us a huge benefit.

@intel/dpcpp-cfe-reviewers Is Slava's idea possible? Can we remove all code for the kernel or even delete the kernel function inside Sema or somewhere else inside the FE such that codegen doesn't generate any IR for the function? My intuition is it's not possible and we would need to modify CodeGenFunction or something but I am not an FE expert. Thanks!

Reply:
Can you clarify what exactly it is that is being removed here? Possibly with an example before and after this change? I wouldn't expect kernel code to be generated during host compilation. Do we just need to add some ESIMD check here - https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp#L6856 ?

Reply:
Idea from Slava: Use the ESIMD metadata instead of the function name, let me try this.

Reply:
Yes, It was the idea from the very beginning, I only did not realize that ESIMD metadata probably was not kept for HOST compilation. Thank you!

Reply:
@v-klochkov I tried this and it was as I expected, we don't add the metadata for the host. However even if I change that, it doesn't fix the problem because the simd* class member functions do not use the ESIMD_FUNCTION/ESIMD_API macros so they never get annotated, and in this case the function we want to remove is the simd_view::read function. I could annotate all of the classes with the macro, but I think the current namespace-based solution is easier to maintain. What do you think?

Reply:
Well, still not clear why it did not work easily. In particular, why if function/kernel F has 'is_esimd' metadata, then not erase it's body on host assuming everything else if removed automatically by DCE (hmm, this approach would not work with -O0 though).
I don't want to hold this patch, which is definitely good for it's purpose - save compile time by eliminating ESIMD code on host.

Reply:
Let me quickly investigate this, it's an interesting question.

Reply:
Ok, I think I got it. It looks like it's a combination of two things.

1) If you don't pass any `-O` flag in clang, by default clang does not run GlobalDCE on the host, so nothing is removed besides the kernel body I delete in my pass.

2) Even if you pass `-O1` or higher and GlobalDCE is run, not all functions are removed by GlobalDCE because of the linkage. In particular the simd_view::read function again is not optimized because it has linkonce_odr, not internal_linkage. This is because those member functions don't have the ESIMD_FUNCTION macro where I set the linkage to internal_linkage if this option is passed.

So again we hit the problem of having to annotate all of these member functions, so I think the current solution is the simplest, albeit a bit hacky.

Reply:
>So I don't think dropping the kernel is going to work because instead we need to keep it but modify the implementation. I don't think we can do that in Clang and instead need to do it in IR but let me know if I'm wrong

From the example you shared, it looks like you just want to remove kernel body in some use-case. That can be done by modifying the visitors in SemaSYCL.  There is a visitor we use to create openCL kernel body, which can just be skipped in this case. It might be more involved work though and I don't know if it is necessarily better to do that vs having this pass.  IMO the headers should be modified as @AlexeySachkov mentioned. This is ok as a temporary workaround but probably shouldn't be final solution to this problem.

Reply:
Thanks for the reply, yeah I think that would work too, the pass is relatively simple and runs early so I don't think there's a huge difference overall. The pass touches other functions besides the kernel too so if we are running DCE deleting the kernel is all we need, but sometimes DCE isn't run (default pipline doesn't seem to run it) and in that case the non-kernel functions would still be around which might cause problems. We will work to improve the headers.